scripts/components/Server.js

```diff
@@ -7,6 +7,10 @@ import chevron from '../../assets/chevron-down-solid.svg'
 
 const RANDOM_SAMPLE_SIZE = 6
 
+function shuffleArray(array) {
+  return array.sort(() => 0.5 - Math.random())
```

Comment:
A frisk and review of this shuffle implementation shows that not only is it inefficient - but *it's not even that random*.

Let's detail why this might be the case:

## Comparisons are non-deterministic and intransitive
The function passed to `array.sort` accepts (or is supposed to accept) an `(a, b)`, and the result returned determines the relationship as follows:
- If the return value is negative, `a < b`, and so `a` comes first.
- If the return value is positive, `a > b`, and so `b` comes first.
- If the return value is zero, `a = b`, and so their order is of no significance.

`array.sort` can operate with such a function because it operates under the assumption of any sortable set: that comparison transitivity applies.

Specifically, in a set of `(1, 2, 3)`, the assertion that both `2 > 1` and `3 > 2` implies that `3 > 1`, even if we do not explicitly check that this is true.

As in this case the output of the comparison function is completely random, the actual output of a sort in this manner is not consistent or logical, and is instead based on which combinations of comparisons the sorter uses to reach its conclusion.

This leads directly into the next point:
## Inconsistent comparisons leads to inconsistent placement assertions
Under the assumption that our comparisons are transitive, there are some comparisons that we don't need to do. We know the most about the things we process first, in this case, being the first element we see in the array.

With that in mind, let's use this shuffler to shuffle a simple array: `[0, 1, 2, 3, 4]`. We'll evaluate it by seeing how often any given element appears **first** in the final, shuffled array, over 200,000,000 iterations:
```js
final = {0: 0, 1: 0, 2: 0, 3: 0, 4: 0}
for (var i = 0; i < 200000000; i++) {
    final[[0, 1, 2, 3, 4].sort(() => 0.5 - Math.random())[0]]++;
}
```
Taking a run of this gives us this kind of result:
```js
{ '0': 61530652,
  '1': 61512931,
  '2': 41012738,
  '3': 23445242,
  '4': 12498437 }
```
![plot](https://user-images.githubusercontent.com/8049998/54499939-fc486d00-495a-11e9-881d-e9e2cc588726.png)

Yikes, that's pretty egregiously biased!

What we're seeing here is a manifestation of the concept I explained above: because `array.sort` assumes the function will obey transitivity, it skips out on doing every possible comparison, as most of the time, it *doesn't need to*.

In an example case, if element 1 evaluates (at a 50% chance) that it comes before element 2, the later 50% chance that element 2 evaluates before element 3 means element 3 will **never contest element 1**, as in normal sorting conditions, there is no need for this to happen.

In addition, `array.sort` covers *neighboring* elements primarily. Element 1 may need only to win the 50/50 once to stay where at slot 1, but element 5 has to beat the coin flip 4 times in a row to get to the same place.

The solution, of course, is to find a solution that is both transitive and deterministic. In fact, we can actually avoid the problem completely by not treating this like a sort at all. The popular [Fisher-Yates shuffle](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle) works by instead applying randomness to *indices* rather than elements, this way, no element has a bias regarding its neighbors as any element can choose to swap with any other element.

Here is an implementation of the Fisher-Yates by [Mike Bostock](https://bost.ocks.org/mike/shuffle/):
```js
function shuffle(array) {
  var m = array.length, t, i;

  // While there remain elements to shuffle…
  while (m) {

    // Pick a remaining element…
    i = Math.floor(Math.random() * m--);

    // And swap it with the current element.
    t = array[m];
    array[m] = array[i];
    array[i] = t;
  }

  return array;
}
```

And here is a collapsed version used by the popular library [lodash](https://github.com/lodash/lodash/):
```js
function baseRandom(lower, upper) {
  return lower + Math.floor(Math.random() * (upper - lower + 1));
}

function shuffleSelf(array, size) {
  var index = -1,
      length = array.length,
      lastIndex = length - 1;

  size = size === undefined ? length : size;
  while (++index < size) {
    var rand = baseRandom(index, lastIndex),
        value = array[rand];

    array[rand] = array[index];
    array[index] = value;
  }
  array.length = size;
  return array;
}
```

It is probably not a good idea to rip these implementations verbatim, but if an appropriate library with shuffle is not to be used, we should make sure we're doing it in a way that *actually* ensures our sample is "truly random".

Reply:
Uh, by "truly random", i meant picking random emojis from the entire server instead of just a small section of the server that never changes (which makes picking random emojis pointless), not literally truly random, and that shuffle implementation is the same one that shuffles the community servers.

Reply:
This algorithm's flaws will mean the end result will almost always have very little variance from the original data due to the interpair swapping rules described above.

For servers with large amounts of emotes, this approach might even be *less* random than the index-and-slice approach, as the most likely outcome is that the array slice used will contain only a small subset of the emotes outside of its range, with the major difference in runs being that the emotes may appear in different orders (e.g. `A C B` rather than `A B C`).